### PR TITLE
Add runtime-extra-platforms pipeline to have matching runtime PR and Rolling builds

### PR DESCRIPTION
**Pull request by safern**
This change makes the `runtime` pipeline be practically the same in between PR Builds and Rolling Builds. I say practically, because the only differences that remain for some platforms is that in PRs we build them using `Debug` but on Rolling we use `Release`. The reason for this is in order to get coverage in `Debug` and `Release` without having to add extra legs to the PRs causing more resources to be used and more time for the builds to complete, since the creation of the `dotnet/runtime` repository, we haven't seen much failures caused by this `Debug` vs `Release` differences, so I think it is worth to keep.

### Motivation

Since the creation of the repository we have been struggling to improve the pass rate on PRs and when improved maintain at a high rate. The reason for this is because on PRs we run tests on a reduced set of platforms due to time constraints and resources, so it is hard to measure a 1:1 pass rate in between PRs and Rolling because in Rolling we run a super set of platforms that might cause the Rolling build to fail, even if the PR matrix is passing, so that doesn't reflects the real pass rate numbers.

In order to make these numbers more transparent we have decided to introduce a new pipeline, `runtime-extra-platforms` which contains the `extra` platforms that we ran tests on Rolling builds only, so essentially this new pipeline structure combined is equal to what we had for `runtime` Rolling build.

Also, one thing that was considered painful was when I did a change that affected an architecture or platform that was tested on Rolling only, I had to push a change to my PR and then manually schedule a build pointing to my branch in order for those to get tested. With this change, we add the ability for contributors to run `/azp run runtime-extra-platforms` as shown in this PR and get the extra platforms tested on the current branch, also, using the `evaluate-changed-paths` infra, so if only changes from one partition are made, only tests that affect that partition will be executed.

As part of this change we are also removing the `runtime-manual` pipeline and moved everything that was there into the new pipeline, and removed from `runtime-staging` the platforms that are already stable and added them to the `extra-platforms` pipeline in order to avoid duplication and save resources..

### Notes

This new pipeline will be treated as an extension of the runtime pipeline and we will measure pass rate on CI Council, will be monitored by the test sheriff and we will also open issues with `blocking-clean-ci` label for this pipeline so that action is taken immediately to unblock the pipeline. We don't intend to treat this pipeline as "optional" and should strive towards a high pass rate as these platforms are important for the product.

### CODE REVIEW

**Comment by lambdageek** [comment]
Is there a summary per-platform/configuration of what coverage is in `runtime` and what's in `runtime-extra-platforms`?

**Comment by safern** [comment]
> Is there a summary per-platform/configuration of what coverage is in runtime and what's in runtime-extra-platforms?

I can get that and put it in the description.

**Comment by lambdageek** [comment]
Also I assume the unstable runtime-staging platforms are staying in staging? no change there?

**Comment by safern** [comment]
> So the stable runtime-staging platforms are graduating to extra. And I assume the unstable runtime-staging platforms are staying in staging? no change there?

Exactly, I left in staging:

- Android_arm64 runtime tests as those have a comment that should be excluded from PRs, @steveisok  is following on that one.
- iOS/TVOs simulators as per @steveisok are not stable yet.
- Android emulators as per @steveisok are more stable than simulators but not stable enough to move them to extra.
- OSX arm64 as I need to follow up to see how stable that one is.

**Comment by lambdageek** [comment]
I chatted with @steveisok on Discord, so I'm going to summarize in my own words what this PR is doing; I hope it's accurate:

1. For rolling builds: `runtime` is slimming down to the same smoketest-like configurations that we have on PRs.  The `runtime-extra-platforms` is picking up the slack.  And we will monitor both `runtime` and `runtime-extra-platforms` for CI health and create issues, etc.  Some stable platforms from `runtime-staging` are graduating to `runtime-extra-platforms`.
2. For Pull Requests: if I want confidence in my pull request, this PR doesn't really change anything.  The runtime pipeline on PRs is more or less the same smoketest-like coverage that hits a limited set of configurations.  But whereas before I had to do `/azp run runtime-manual` to get more coverage, now I do `/azp run runtime-extra-platforms`.  And the bonus is that unlike runtime-manual this will cover _exactly_ (modulo changed paths) what the rolling build's `runtime-extra-platforms` will run later that night.

**Comment by marek-safar** [comment]
Just nitpicking on the naming ;-) If I understood the intention this is not about platforms nor about anything that is more than necessary. Could we name it in a way that it survives more than one infrastructure clean up?

**Comment by safern** [comment]
That's perfectly accurate. Thanks @lambdageek

**Comment by safern** [comment]
>  Could we name it in a way that it survives more than one infrastructure clean up?

Sure 😀. Any suggestions that come to mind?

I can think of: 
runtime-extra-platforms
runtime-extra-configurations
runtime-extended
runtime-extra

**Comment by trylek** [comment]
I think Marek's point was the it's actually not anything 'extra', it's rather something that's 'missing' in the default PR testing, so it might be more descriptive to base the naming on something like 'runtime-remaining-platforms', 'runtime-non-pr-platforms' or 'runtime-complementary-platforms'. I'm not emotionally attached to any of these variants, just trying to brainstorm.

**Comment by trylek** [comment]
Not sure to what extent it would be precise but something like 'runtime-mono-platforms' or 'runtime-device-platforms' might also capture the gist of the distinction.

**Comment by steveisok** [comment]
> Not sure to what extent it would be precise but something like 'runtime-mono-platforms' or 'runtime-device-platforms' might also capture the gist of the distinction.

Naming is tough.  It's definitely not mono specific and although most are device oriented, that's not absolute (wasm aot, for example).  Is there a strong enough name ;-)?

File: eng/pipelines/runtime.yml [hunk 1]
```diff
@@ -460,78 +459,6 @@ jobs:
...
-      # extra steps, run tests
-      extraStepsTemplate: /eng/pipelines/common/templates/runtimes/wasm-runtime-and-send-to-helix.yml
-      extraStepsParameters:
-        creator: dotnet-bot
-        testRunNamePrefixSuffix: Mono_$(_BuildConfig)
-
-#
-# Build the whole product using Mono for Android and run runtime tests with Android emulator
-#
-- template: /eng/pipelines/common/platform-matrix.yml
```
Comment (fanyang-mono) [inline eng/pipelines/runtime.yml:502]: Was it moved to runtime-extra-platforms pipeline? I didn't see it there.

File: eng/pipelines/runtime-staging.yml [hunk 2]
```diff
@@ -438,8 +218,7 @@ jobs:
         or(
           eq(dependencies.evaluate_paths.outputs['SetPathVars_runtimetests.containsChange'], true),
           eq(dependencies.evaluate_paths.outputs['SetPathVars_mono.containsChange'], true),
-          eq(variables['isManualOrIsNotPR'], true),
-          eq(variables['isFullMatrix'], true))
+          eq(variables['isRollingBuild'], true))
       ${{ if eq(variables['isManualOrIsNotPR'], true) }}:
```
Comment (fanyang-mono) [inline eng/pipelines/runtime-staging.yml:222]: You missed this line.

File: eng/pipelines/runtime-staging.yml [hunk 3]
```diff
@@ -321,49 +144,7 @@ jobs:
...
-        condition: >-
-          or(
-          eq(variables['librariesContainsChange'], true),
-          eq(variables['monoContainsChange'], true),
-          eq(variables['isManualOrIsNotPR'], true),
-          eq(variables['isFullMatrix'], true))
+          eq(variables['isRollingBuild'], true))
 
 #
 # Build the whole product using Mono for Android and run runtime tests with interpreter
```
Comment (fanyang-mono) [inline eng/pipelines/runtime-staging.yml:150]: This leg is quite stable as well and could be moved to `runtime-extra-platforms`

**Comment by jkotas** [comment]
Is this also going to include outerloop tests, or are outerloop tests a separate variable?

My naming suggestion would `runtime-full` or `runtime-full-matrix`

**Review by fanyang-mono — COMMENTED** [review]
Based on my understanding, I am proposing the name of the new pipeline to be `runtime-rolling-only`. And how to trigger the rolling build only legs belonging to `runtime-staging`? I see there are still a large amount of rolling build only lanes. Why don't we move all of them to this new pipeline?

**Comment by safern** [comment]
This is definitely not mono specific, we are also running libraries tests on regular configurations; I.e linux_x64 but we run tests on more platforms like OpenSuse or Fedora, etc.

I think I like `runtime-complementary-platforms` the best but I think `runtime-rolling-only`  also meets what the intent of the pipeline is.

For `runtime-full-matrix` we wanted to avoid `full` in the naming as it is not a full run per sei as it doesn't include what's covered by the `runtime` pipeline.

**Comment by steveisok** [comment]
> > No, this is only innerloop tests but on platforms/configurations/runtime variants that we don't cover on PR
> 
> How should people decide whether they want to the extended run or outer loop run or both?

The easiest example of why you would decide to run extended is when adding a test to, say, wasm.  On PR's, I would have no automatic way of validating the change for all wasm configurations and would have to manually trigger the extended pipeline.

**Comment by trylek** [comment]
As a side note, it looks like this definitely merits precise and easy to find documentation. And the documentation should also clearly state that it's not the case that <code>outerloop</code> is a magical superset of all of these, in fact current <code>runtime-coreclr outerloop</code> pipeline doesn't exercise Mono at all and in general we're not running Pri1 tests on Mono these days (historically because making even the Pri0 tests pass was quite a bit of a challenge, currently due to HW limitations).

**Comment by safern** [comment]
Good call. Ill include documentation on the PR.

File: eng/pipelines/runtime.yml (hunk 1 above)
Comment (safern) [reply eng/pipelines/runtime.yml:502]: Good catch. For some reason I thought I moved it. I will include it in my next commit.

File: eng/pipelines/runtime-staging.yml (hunk 2 above)
Comment (safern) [reply eng/pipelines/runtime-staging.yml:222]: Thanks.

File: eng/pipelines/runtime-staging.yml (hunk 3 above)
Comment (safern) [reply eng/pipelines/runtime-staging.yml:150]: Are we sure it is stable enough to move it?

**Comment by safern** [comment]
Alright, sorry for the spam.

I added docs and addressed PR Feedback. This is now ready for another round of reviews.

File: eng/pipelines/runtime-staging.yml (hunk 3 above)
Comment (fanyang-mono) [reply eng/pipelines/runtime-staging.yml:150]: Yes, it has been green for a while (more than 2 weeks).

**Comment by fanyang-mono** [comment]
LGTM!

File: eng/pipelines/runtime-staging.yml (hunk 3 above)
Comment (safern) [reply eng/pipelines/runtime-staging.yml:150]: Ok, I will address this on a separate PR to not reset CI.

Comment (safern) [reply eng/pipelines/runtime-staging.yml:150]: I need to push a commit so I'll include it.

**Review by jkoritzinsky — APPROVED** [review]

**Review by trylek — APPROVED** [review]
Looks great to me, thank you!

File: eng/pipelines/common/variables.yml [hunk 4]
```diff
@@ -11,19 +11,17 @@ variables:
...
+- name: isExtraPlatformsBuild
+  value: ${{ eq(variables['Build.DefinitionName'], 'runtime-extra-platforms') }}
+- name: isNotExtraPlatformsBuild
+  value: ${{ ne(variables['Build.DefinitionName'], 'runtime-extra-platforms') }}
 
 # We only run evaluate paths on runtime, runtime-staging and runtime-community pipelines on PRs
 # keep in sync with /eng/pipelines/common/xplat-setup.yml
 - name: dependOnEvaluatePaths
-  value: ${{ and(eq(variables['Build.Reason'], 'PullRequest'), in(variables['Build.DefinitionName'], 'runtime', 'runtime-staging', 'runtime-community')) }}
+  value: ${{ and(eq(variables['Build.Reason'], 'PullRequest'), in(variables['Build.DefinitionName'], 'runtime', 'runtime-staging', 'runtime-community', 'runtime-extended-platforms')) }}
```
[the remark this answers is not in the document]
Comment (safern) [reply eng/pipelines/common/variables.yml]: It was the original name, we renamed the pipeline and I forgot to update the yml. However I updated it, seems like in between your review, so this is resolved 😄

**Comment by safern** [comment]
Test failure was disabled on: https://github.com/dotnet/runtime/pull/64006

**Pull request merged**